### DISCUSSION TITLE
Installing R Package on Mox

### DISCUSSION THREAD

**Discussion by aspencoyle** [RobertsLab/resources#1218] [Q&A] [answered]
Hi! 

I'm trying to install a package (specifically pheatmap) onto Mox. I'm trying to use it in the manual module creation script that Laura made - my computer just doesn't have the memory for it.

I loaded R version 3.6.0, ran `install.packages("pheatmap")`, and then started up the interactive RStudio session and ran `library(pheatmap, lib.loc = "/gscratch/srlab/afcoyle/rpackages/")`

This returned `Error: package or namespace load failed for ‘pheatmap’:
 package ‘pheatmap’ was installed before R 4.0.0: please re-install it`
 
 I then tried to re-download the package using R version 4.0.5 (as described in [Yaamini's discussion post here](https://github.com/RobertsLab/resources/discussions/1178)). However, when trying to install using `install.packages("pheatmap", lib="gscratch/srlab/afcoyle/rpackages"), it returned 

```Warning in install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") : lib = "gscratch/srlab/afcoyle/rpackages"' is not writable```

I ignored and it returned

```
The downloaded source packages are in
        '/tmp/RtmpePJGqF/downloaded_packages'
Warning messages:
1: In install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") :
  installation of package 'rlang' had non-zero exit status
2: In install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") :
  installation of package 'lifecycle' had non-zero exit status
3: In install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") :
  installation of package 'scales' had non-zero exit status
4: In install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") :
  installation of package 'pheatmap' had non-zero exit status
  ```
  
When I then tried to load using `library(pheatmap, lib.loc = "/tmp/RtmpePJGqF/downloaded_packages")` it just returned `Error in library(pheatmap, lib.loc = "/tmp/RtmpePJGqF/downloaded_packages") : 
  no library trees found in 'lib.loc'`
  
  Any advice for how to load this dang package?

**Comment by kubu4** [comment]
At a minimum, you'll have to install all those other packages that returned errors first, then install pheatmap.

**Reply by aspencoyle** [reply]
I tried installing those individually, and couldn't install any of them either - all returned some variation of `Warning message:
In install.packages("rlang", lib = "/gscratch/srlab/afcoyle/rpackages") :
  installation of package 'rlang' had non-zero exit status`

**Comment by kubu4** [comment, marked as answer]
Didn't notice this part before:

```R
Warning in install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") : lib = "gscratch/srlab/afcoyle/rpackages"' is not writable
```

Try changing permissions on that directory so it's writable.

**Reply by aspencoyle** [reply]
Alright, reviewed this and some good news, some bad.

Good news: I made a really obvious mistake  - my R packages directory is called "R_libs_singularity", not "rpackages", so I was trying to install packages to a directory that doesn't exist. So the `Warning in install.packages("pheatmap", lib = "gscratch/srlab/afcoyle/rpackages") : lib = "gscratch/srlab/afcoyle/rpackages"' is not writable` warning message was caused by that.

Bad news: When fixed, I'm still getting variations of `Warning message: In install.packages("rlang", lib = "/gscratch/srlab/afcoyle/rpackages") : installation of package 'rlang' had non-zero exit status`, regardless of whether I'm trying to install `pheatmap` or its dependencies

One thing that I didn't specify earlier is that I'm using the R installation that Yaamini installed a few weeks ago in /gscratch/srlab/yaaminiv/Rinstall, as I was having problems when trying to replicate that installation. Hopefully that's all OK and isn't causing any issues, but just thought I'd mention it in case!

**Reply by kubu4** [reply]
What is the full error message when you try to install `rlang`?

**Reply by aspencoyle** [reply]
```
install.packages("rlang", lib = "/gscratch/srlab/afcoyle/R_libs_singularity")
trying URL 'https://ftp.osuosl.org/pub/cran/src/contrib/rlang_0.4.11.tar.gz'
Content type 'application/x-gzip' length 861727 bytes (841 KB)
==================================================
downloaded 841 KB

During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C"
2: Setting LC_TIME failed, using "C"
3: Setting LC_MESSAGES failed, using "C"
4: Setting LC_MONETARY failed, using "C"
5: Setting LC_PAPER failed, using "C"
6: Setting LC_MEASUREMENT failed, using "C"
* installing *source* package 'rlang' ...
** package 'rlang' successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/gscratch/srlab/yaaminiv/Rinstall/lib64/R/include" -DNDEBUG -I./lib/  -I/usr/local/include  -fvisibility=hidden -fpic  -g -O2  -c capture.c -o capture.o
gcc -I"/gscratch/srlab/yaaminiv/Rinstall/lib64/R/include" -DNDEBUG -I./lib/  -I/usr/local/include  -fvisibility=hidden -fpic  -g -O2  -c export.c -o export.o
In file included from export.c:1:0:
export/exported.c: In function 'rlang_env_bind_list':
export/exported.c:93:3: error: 'for' loop initial declarations are only allowed in C99 mode
   for (r_ssize i = 0; i < n; ++i) {
   ^
export/exported.c:93:3: note: use option -std=c99 or -std=gnu99 to compile your code
In file included from export.c:1:0:
export/exported.c: In function 'rlang_is_string':
export/exported.c:572:3: error: 'for' loop initial declarations are only allowed in C99 mode
   for (r_ssize i = 0; i < n; ++i) {
   ^
make: *** [export.o] Error 1
ERROR: compilation failed for package 'rlang'
* removing '/gscratch/srlab/afcoyle/R_libs_singularity/rlang'

The downloaded source packages are in
        '/tmp/RtmpyFCkLv/downloaded_packages'
Warning message:
In install.packages("rlang", lib = "/gscratch/srlab/afcoyle/R_libs_singularity") :
  installation of package 'rlang' had non-zero exit status
```

**Reply by kubu4** [reply]
Thanks. It's a compiler error. You'll probably need to load a compiler module before trying again. Use `module avail` to list all modules. See if you can find one that looks like it will match the `gnu99` stuff...

I'll try to look in a bit to provide more guidance.

**Reply by aspencoyle** [reply]
Gotcha, thanks! I just checked, and can't see any modules that match anything close to gnu99.

Also just to give some background, I'm trying to load this package so I can use it in Rstudio on Mox , and remember that early on, you installed the DESeq2 and WGCNA packages for me for that purpose. If that's different from what I've been doing, can you share the method you used to do that?

**Reply by kubu4** [reply]
Yes, you should be updating (or building a new) Singularity container! Instructions are in Handbook.

**Reply by yaaminiv** [reply]
> Yes, you should be updating (or building a new) Singularity container! Instructions are in Handbook.

FYI I wanted to do something similar to install R packages `bsseq` and `DSS` for use on the R Studio server, and building a new Singularity container worked for me!

**Reply by aspencoyle** [reply]
I did that and it worked perfectly - thanks so much!
